tower-balance/src/p2c/service.rs

```diff
@@ -106,120 +177,106 @@ impl<D: Discover> Balance<D> {
-                        (Async::NotReady, new_bidx)
+                        (None, new_bidx)
                     }
                 };
 
-                let (bload, aidx) = match self.poll_endpoint_index_load(bidx) {
-                    Ok(ready) => (ready, aidx),
-                    Err(e) => {
-                        info!("evicting failed endpoint: {}", e.into());
-                        let _ = self.endpoints.swap_remove_index(bidx);
-                        let new_aidx = Self::repair_index(aidx, bidx, self.endpoints.len())
+                let (bload, aidx) = match self.poll_ready_index_load(bidx) {
+                    Some(bload) => (Some(bload), aidx),
+                    None => {
+                        let new_aidx = Self::repair_index(aidx, bidx, self.ready.len())
                             .expect("random indices must be distinct");
-                        (Async::NotReady, new_aidx)
+                        (None, new_aidx)
                     }
                 };
 
                 trace!("load[{}]={:?}; load[{}]={:?}", aidx, aload, bidx, bload);
 
                 let ready = match (aload, bload) {
-                    (Async::Ready(aload), Async::Ready(bload)) => {
+                    (Some(aload), Some(bload)) => {
                         if aload <= bload {
-                            Async::Ready(aidx)
+                            Some(aidx)
                         } else {
-                            Async::Ready(bidx)
+                            Some(bidx)
                         }
                     }
-                    (Async::Ready(_), Async::NotReady) => Async::Ready(aidx),
-                    (Async::NotReady, Async::Ready(_)) => Async::Ready(bidx),
-                    (Async::NotReady, Async::NotReady) => Async::NotReady,
+                    (a, b) => a.map(|_| aidx).or_else(|| b.map(|_| bidx)),
                 };
                 trace!(" -> ready={:?}", ready);
-                Ok(ready)
+                ready
             }
+        };
+
+        if self.next_ready_index.is_some() {
+            return Ok(Async::Ready(()));
         }
+
+        Ok(Async::NotReady)
     }
 
-    /// Accesses an endpoint by index and, if it is ready, returns its current load.
-    fn poll_endpoint_index_load<Svc, Request>(
-        &mut self,
-        index: usize,
-    ) -> Poll<Svc::Metric, Svc::Error>
-    where
-        D: Discover<Service = Svc>,
-        Svc: Service<Request> + Load,
-        Svc::Error: Into<error::Error>,
-    {
-        let (_, svc) = self.endpoints.get_index_mut(index).expect("invalid index");
-        try_ready!(svc.poll_ready());
-        Ok(Async::Ready(svc.load()))
+    /// Accesses a ready endpoint by index and returns its current load.
+    fn poll_ready_index_load(&mut self, index: usize) -> Option<<D::Service as Load>::Metric> {
+        let (_, svc) = self.ready.get_index_mut(index).expect("invalid index");
+        let load = match svc.poll_ready() {
```

Comment:
I wonder if it's even necessary to `poll_ready` here. Doesn't the `Service` only end up in `self.ready` if it has already returned `Ready` from `poll_ready`?

Reply:
I commented on this below, but I now realize (I think) that this is due to cancellations. Explaining that in a comment would be helpful.